### PR TITLE
Implemented DISTINCT ON in SQLServer

### LINKED ISSUE #5840: Wrong SQL Syntax in Bookmarker Tutorial (CakePHP 3)
Hey,

I have tested the CakePHP 3 Bookmarker Tutorial with SqlSRV 3.1 and SQL Server 2012. When searching for tags, the generated SQL is 
```sql
SELECT DISTINCT ON (Bookmarks.id, Bookmarks.title, Bookmarks.url) Bookmarks.id AS [Bookmarks__id], 
Bookmarks.user_id AS [Bookmarks__user_id], 
Bookmarks.title AS [Bookmarks__title], 
Bookmarks.description AS [Bookmarks__description], 
Bookmarks.url AS [Bookmarks__url], 
Bookmarks.created AS [Bookmarks__created], 
Bookmarks.updated AS [Bookmarks__updated], 
Tags.id AS [Tags__id], 
Tags.title AS [Tags__title], 
Tags.created AS [Tags__created], 
Tags.updated AS [Tags__updated], 
BookmarksTags.bookmark_id AS [BookmarksTags__bookmark_id], 
BookmarksTags.tag_id AS [BookmarksTags__tag_id]
FROM bookmarks Bookmarks
INNER JOIN tags Tags ON Tags.title in (:c0) 
INNER JOIN bookmarks_tags BookmarksTags ON (Bookmarks.id = (BookmarksTags.bookmark_id) AND Tags.id = (BookmarksTags.tag_id))
```

Best Regards

### PR DESCRIPTION
**Pull request by lorenzo**
Some recent refactoring made this feature works as it does for Mysql
and Sqlite.

Closes #5840

### CODE REVIEW

**Comment by lorenzo** [comment]
So why does it pass on my windows machine...?

**Comment by lorenzo** [comment]
Investigating :P

**Comment by markstory** [comment]
Could it be 2008 vs 2012?

**Comment by lorenzo** [comment]
Found the problem, I was running tests with sqlite. This feature needed a real implementation, which I did :D

File: src/Database/Dialect/SqlserverDialectTrait.php [hunk 1]
```diff
@@ -132,6 +132,57 @@ protected function _pagingSubquery($original, $limit, $offset)
...
+     * @param Query $original The query to be transformed
+     * @return Query
+     */
+    protected function _transformDistinct($original)
+    {
+        if (!is_array($original->clause('distinct'))) {
+            return $original;
+        }
+
+        $query = clone $original;
```
Comment (markstory) [inline src/Database/Dialect/SqlserverDialectTrait.php]: This is going to make a shallow clone. Is that ok?

File: src/Database/Dialect/SqlserverDialectTrait.php [hunk 2]
```diff
@@ -132,6 +132,57 @@ protected function _pagingSubquery($original, $limit, $offset)
...
+                ];
+            })
+            ->limit(null)
+            ->offset(null)
+            ->order([], true);
+
+        $outer = new Query($query->connection());
+        $outer->select('*')
+            ->from(['_cake_distinct_' => $query])
+            ->where(['_cake_distinct_pivot_' => 1]);
```
Comment (markstory) [inline src/Database/Dialect/SqlserverDialectTrait.php]: Fancy stuff.

File: src/Database/Dialect/SqlserverDialectTrait.php (hunk 1 above)
Comment (lorenzo) [inline src/Database/Dialect/SqlserverDialectTrait.php]: Yes, that is fine

**Comment by lorenzo** [comment]
Merging as this didn't break any additional tests

**Pull request merged**